autocomplete.ts

```diff
@@ -0,0 +1,127 @@
+import CodeMirror from "codemirror";
+import { BasicREPL } from "./repl";
+
+export function populateAutoCompleteSrc(repl: BasicREPL): Array<any> {
+    var defList: string[] = [];
+    var classMethodList: string[] = [];
+    //get variable names for autocomplete
+    repl.currentTypeEnv.globals.forEach((val, key) => {
+        //don't add functions into variable list
+        // if (val.tag != "callable") {
+        //   defList.push(key);
+        // }
+    });
+    //get class names for autocomplete
+    repl.currentTypeEnv.classes.forEach((val, key) => {
+        defList.push(key);
+        //second element denotes class methods
+        if (val.length > 1) {
+            val[1].forEach((v, k) => {
+                classMethodList.push(k + "()");
+            });
+        }
+    });
+    //get function names for autocomplete
+    repl.currentTypeEnv.functions.forEach((val, key) => {
+        defList.push(key + "()");
+    });
+    return [defList, classMethodList];
+}
+
+export function autocompleteHint(editor: any, keywords: string[], getToken: any) {
+    // Find the token at the cursor
+    var currPos = editor.getCursor();
+    var token = getToken(editor, currPos),
+        tprop = token;
+    var isClassMethod = false;
+    if (token.string[token.string.length - 1] === ".") {
+        isClassMethod = true;
+        token = tprop = {
+            start: currPos.ch,
+            end: currPos.ch,
+            string: "",
+            state: token.state,
+            type: "property",
+        };
+    } else if (token.type == "property") {
+        isClassMethod = true;
+    }
+    //ignore any non word or property token
+    else if (!/^[\w$_]*$/.test(token.string)) {
+        token = tprop = {
+            start: currPos.ch,
+            end: currPos.ch,
+            string: "",
+            state: token.state,
+            className: token.string == ":" ? "python-type" : null,
+        };
+    }
+
+    if (!context || isClassMethod) {
+        var context = [];
+        context.push(tprop);
+
+        var completionList = getCompletions(keywords, token, context);
+        completionList = completionList.sort();
+        //show dropdown with one word (restrict automatic autocomplete of single word)
+        if (completionList.length == 1) {
+            completionList.push(" ");
+        }
+    }
+    return {
+        list: completionList,
+        from: CodeMirror.Pos(currPos.line, token.start),
+        to: CodeMirror.Pos(currPos.line, token.end),
+    };
+}
+
+function getCompletions(wordList: any, token: any, context: any) {
+    var completions: any[] = [];
+    var prefix = token.string;
+
+    if (context) {
+        // If this is a property, see if it belongs to some object we can
+        // find in the current environment.
+        var obj = context.pop(),
+            base;
+
+        if (obj.type == "variable") {
+            base = obj.string;
+        } else if (obj.type == "variable-3") {
```

Comment:
whoa, I have no idea what "variable-3" is.

Reply:
Thanks, I would update